File: model_comparison/compare_with_Kay2013.py
# compare_with_Kay2013.py
#
# this script runs the SCO model on the bandpassed stimuli from http://kendrickkay.net/socmodel/,
# then exports the parameters so the MATLAB version of the model (from the same url) can be run on
# the same stimuli with the same parameters. We then compare the predicted responses between the
# two models.
#
# By William F. Broderick

import sco
import neuropythy.freesurfer as nfs
import os
import glob
import imghdr
import scipy.io as sio
import numpy as np


def main(image_base_path, stimuli_idx, subject='test-sub',
         subject_dir='/home/billbrod/Documents/SCO-test-data/Freesurfer_subjects'):
    """
    Run python SCO and Matlab SOC on the same batch of images

    Arguments
    ---------

    image_base_path: string. This is assumed to either be a directory, in which case the model will
    be run on every image in the directory, or a .mat file containing the image stimuli, in which
    case they will be loaded in and used.

    subject_dir: string or None. If not None, will add this to neuropythy.freesurfer's subject
    paths

    subject: string. The specific subject to run on.
    """
    if subject_dir and subject_dir not in nfs.subject_paths():
        nfs.add_subject_path(subject_dir)
    # if there's just one value and not a list
    if not hasattr(stimuli_idx, '__iter__'):
        stimuli_idx = [stimuli_idx]
    if os.path.isdir(image_base_path):
        # Interestingly enough, this works regardless of whether image_base_path ends in os.sep or
        # not.
        stimulus_image_filenames = glob.glob(image_base_path + os.sep + "*")
        # imghdr.what(img) returns something if img is an image file (it returns a stirng
        # specifying whta type of image it is). If it's not an image file, it returns None.
        stimulus_image_filenames = [img for img in stimulus_image_filenames
                                    if imghdr.what(img)]
        kwargs = {'stimulus_image_filenames': stimulus_image_filenames[stimuli_idx]}
        # and we use the default sco_chain
        sco_chain = sco.sco_chain
    # if it's a .mat file
    elif os.path.splitext(image_base_path)[1] == ".mat":
        # in this case, we assume it's stimuli.mat from
        # http://kendrickkay.net/socmodel/index.html#contentsofstimuli. We want to grab only the
        # 'images' key from this and only images 226 through 260/end (stimulus set 3; note that we
        # have to convert from MATLAB's 1-indexing to python's 0-indexing), since each entry is a
        # single frame and thus easy to handle.
        stimulus_images = sio.loadmat(image_base_path)
        stimulus_images = stimulus_images['images'][0, stimuli_idx]
        # some of the stimuli have multiple frames associated with them; they're just variations on
        # the same image, so we only take one of them to make predictions for.
        stimulus_images = np.asarray([im if len(im.shape)==2 else im[:, :, 0] for im in stimulus_images])
        kwargs = {'stimulus_images': stimulus_images}
        # in this case, we already have the stimulus images, so we don't need the sco chain to do
        # the importing of them.
        # We need to modify the stimulus chain that's part of sco_chain because we don't need the
        # import_stimulus_images step.
        stim_chain = (
            ('calc_stimulus_default_parameters', sco.stimulus.core.calc_stimulus_default_parameters),
            ('calc_normalized_stimulus', sco.stimulus.core.calc_normalized_stimulus_images))
        # This is our modified chain.
        sco_chain = (('calc_anatomy', sco.anatomy.calc_anatomy),
                     # need to call calc_chain on this to make it ready to go.
                     ('calc_stimulus', sco.core.calc_chain(stim_chain)),
                     ('calc_contrast', sco.contrast.calc_contrast),
                     ('calc_pRF', sco.pRF.calc_pRF),
                     ('calc_normalization', sco.normalization.calc_normalization))
    else:
        raise Exception("Don't know how to handle image_base_path %s, must be directory or .mat "
                        "file" % image_base_path)
    # This prepares the sco_chain, making it a callable object
    sco_chain = sco.core.calc_chain(sco_chain)

    # in order to handle the fact that the Kay2013 matlab code only deals with spatial orientation
    # of 3 cpd, we have to define a new pRF_frequency_preference_function to replace the default.
    def freq_pref(e, s, l):
        # This takes in the eccentricity, size, and area, but we don't use any of them, since we
        # just want to use 3 cpd and ignore everything else. And this must be floats.
        return {3: 1}
    # And this runs it. To make sure it has the same size as the the images used in Kendrick's
    # code, we set the normalized_stimulus_aperture, normalized_aperture_edge_width, and
    # normalized_pixels_per_degree values. We want our final image to be 90x90, with the edge
    # taking up 10% of the total image (ie 5% of the radius). setting pRF_blob_std to 1 should
    # speed things up
    results = sco_chain(subject=subject, max_eccentricity=2.7, normalized_stimulus_aperture=15*2.727,
                        normalized_pixels_per_degree=15, stimulus_aperture_edge_width=15*(3-2.727),
                        pRF_frequency_preference_function=freq_pref, pRF_blob_stds=1, **kwargs)
    return results


# if __name__ == '__main__':
#     import argparse
#     import core
#     parser = argparse.ArgumentParser()
#     parser.add_argument(
#         "image_base_path",
#         help=("string. This is assumed to either be a directory, in which case the model will be "
#               "be run on every image in the directory, or a .mat file containing the image stimuli"
#               ", in which case they will be loaded in and used."))
#     parser.add_argument("subject", help=("string. The specific subject to run on.If unset, will "
#                                          "use test-sub"))
#     parser.add_argument("model_df_path", help=("string. Absolute path to save the model dataframe"
#                                                " at."))
#     parser.add_argument("stimuli_idx", nargs='+', type=int,
#                         help="list of ints. Which indices in the stimuli to run.")
#     parser.add_argument("-s", "--subject_dir", help=("string (optional). If specified, will add to"
#                                                      "neuropythy.freesurfer's subject paths"),
#                         default=None)
#     args = parser.parse_args()
#     if args.subject_dir is not None:
#         results = main(args.image_base_path, np.asarray(args.stimuli_idx), args.subject)
#     else:
#         results = main(args.image_base_path, np.asarray(args.stimuli_idx), args.subject,
#                        args.subject_dir)
#     # for image_names, we use stimuli_idx plus 1, since we use stimuli_idx in python but we'll use
#     # the names in matlab.
#     core.create_model_dataframe(results, np.array(args.stimuli_idx)+1, args.model_df_path)

File: model_comparison/compareWithKay2013.m
% compare_with_Kay2013.m
% 
% this matlab script is the complement of compare_with_Kay2013.py
% and will be run after it to produce results to compare to the
% outputs of the python code.
%
% Assumes you used the default gabor filter options in the python
% code (8 orientations, etc) -- unclear about cycles per image.
% 
% currently, this is the naive way of implementing this, we don't
% take advantage of parallelizing in any way, since this is only
% for small tests.
% 
% By William F. Broderick

function compareWithKay2013(knkutilsPath, stimuliPath, stimuliIdx, voxelIdx, modelPath)
% arguments:
% 
% knkutilsPath: string, path to your knkutils folder, which we'll
% add to the matlab path
% 
% stimuliPath: string, path to stimuli.mat which we'll grab the
% input images from
% 
% stimuliIdx: vector, subset of images to use for these models
% 
% voxelIdx: vector, subset of voxels to generate predictions
% for. There's no reason to do this for every voxel.
% 
% modelPath: string, path to the python dataframe / matlab table
% containing all the model parameters
    modelTable = readtable(modelPath);
    addpath(genpath(knkutilsPath));
    load(stimuliPath, 'images');
    stimuli = images(stimuliIdx);
    clear images;
    
    for imgNum=stimuliIdx
        eval(sprintf('modelTable.MATLAB_predicted_responses_image_%d=nan(height(modelTable),1);',imgNum));
    end
    
    [stimuli, normAlreadyDoneFlag] = preprocessStimuli(stimuli, modelTable);
    
    for vox=voxelIdx
        if ~normAlreadyDoneFlag
            stimuliTmp = divisiveNormalization(stimuli, modelTable, voxelIdx);
        else
            stimuliTmp = stimuli;
        end
        modelTable = generateVoxelPredictions(stimuliTmp, modelTable, vox, stimuliIdx);
    end
    writetable(modelTable,'Kay2013.csv');
end

function [stimuli, normAlreadyDoneFlag] = preprocessStimuli(stimuli, modelTable)
% Preprocess the stimuli for use by all the voxels: resize, make sure
% that all values lay between 0 and 254, then rescale to being between
% 0 and 1 and then mean-center (so they lie between -.5 and .5). Then
% pad stimuli with zeros to reduce edge effects. We then move onto the
% more specialized preprocessing: applying the gabor filters (note
% that this assumes you used the defaults in the python code!) and, if
% we can, divisive normalization (we apply divisive normalization here
% if all voxels have the same divisive normalization parameters)

    % resize the stimuli to 150 x 150 to reduce computational time.
    % use single-format to save memory.
    temp = zeros(150,150,size(stimuli,2),'single');
    for p=1:size(stimuli,2)
        stim = stimuli{:, p};
        if length(size(stim)) == 2
            temp(:,:,p) = imresize(single(stim),[150 150],'cubic');
        else
            % most of the stimuli are made up of multiple frames, but they're all
            % variants on the same thing. For these, we just grab the
            % first frame and ignore the rest because we just want to
            % get a sense of what the model is doing.
            temp(:,:,p) = imresize(single(stim(:,:,1)),[150 150],'cubic');
        end
    end
    stimuli = temp;
    clear temp;

    % ensure that all values are between 0 and 254.
    % rescale values to the range [0,1].
    % subtract off the background luminance (0.5).
    % after these steps, all pixel values will be in the 
    % range [-.5,.5] with the background corresponding to 0.
    stimuli(stimuli < 0) = 0;
    stimuli(stimuli > 254) = 254;
    stimuli = stimuli/254 - 0.5;

    % pad the stimuli with zeros (to reduce edge effects).
    % the new resolution is 180 x 180 (15-pixel padding on each side).
    stimuli = placematrix(zeros(180,180,size(stimuli,3),'single'),stimuli);
   
    % apply Gabor filters to the stimuli.  filters occur at different positions, 
    % orientations, and phases.  there are several parameters that govern the
    % design of the filters:
    %   the number of cycles per image is 37.5*(180/150)
    %   the spatial frequency bandwidth of the filters is 1 octave
    %   the separation of adjacent filters is 1 std dev of the Gaussian envelopes
    %     (this results in a 90 x 90 grid of positions)
    %   filters occur at 8 orientations
    %   filters occur at 2 phases (between 0 and pi)
    %   the Gaussian envelopes are thresholded at .01
    %   filters are scaled to have an equivalent Michelson contrast of 1
    %   the dot-product between each filter and each stimulus is computed
    % after this step, stimulus is images x phases*orientations*positions.
    stimuli = applymultiscalegaborfilters(reshape(stimuli,180*180,[])', ...
                                          37.5*(180/150),-1,1,8,2,.01,2,0);
    
    % compute the square root of the sum of the squares of the outputs of 
    % quadrature-phase filter pairs (this is the standard complex-cell energy model).
    % after this step, stimulus is images x orientations*positions.
    stimuli = sqrt(blob(stimuli.^2,2,2));

    if range(modelTable.Kay2013_normalization_r)==0 && range(modelTable.Kay2013_normalization_s)==0
        % if they all have the same r and s, we can do this here.
        
        normAlreadyDoneFlag = true;
        
        % We call divisive normalization, using parameter values
        % from the first voxel (since they're all the same)
        stimuli = divisiveNormalization(stimuli, modelTable, 1);
    else
        normAlreadyDoneFlag = false;
    end
    
end

function stimuli =  divisiveNormalization(stimuli, modelTable, voxelIdx)
% compute the population term in the divisive-normalization equation.  
% this term is simply the average across the complex-cell outputs 
% at each position (averaging across orientation).
    stimuliPOP = blob(stimuli,2,8)/8;

    % repeat the population term for each of the orientations
    stimuliPOP = upsamplematrix(stimuliPOP,8,2,[],'nearest');
    
    % We only do this if all voxels have the same r and s
    r = modelTable.Kay2013_normalization_r(voxelIdx);
    s = modelTable.Kay2013_normalization_s(voxelIdx);
    
    % apply divisive normalization to the complex-cell outputs.  there are two parameters
    % that influence this operation: an exponent term (r) and a semi-saturation term (s).  
    % the parameter values specified here were determined through a separate fitting 
    % procedure (see paper for details).  for the purposes of this script, we will 
    % simply hard-code the parameter values here and not worry about attempting to fit 
    % the parameters.
    stimuli = stimuli.^r ./ (s.^r + stimuliPOP.^r);
    clear stimuliPOP;

    % sum across orientation.  after this step, stimuli is images x positions.
    stimuli = blob(stimuli,2,8);

end

function modelTable = generateVoxelPredictions(stimuli, modelTable, voxelIdx, stimuliIdx)
% The parameters are [R C S G N C] where
%   R is the row index of the center of the 2D Gaussian (pRF_pixel_centers_image_###_dim0)
%   C is the column index of the center of the 2D Gaussian (pRF_pixel_centers_image_###_dim1)
%   S is the standard deviation of the 2D Gaussian (pRF_pixel_sizes)
%   G is a gain parameter (Kay2013_response_gain)
%   N is the exponent of the power-law nonlinearity (Kay2013_output_nonlinearity)
%   C is a parameter that controls the strength of second-order contrast (Kay2013_SOC_constant)

    % resolution of the pre-processed stimuli; this is taken from
    % Kendrick's socmodel_example, because after the gabor filters
    % are applied, your stimuli are all 90x90
    res = 90;
    % issue a dummy call to makegaussian2d.m to pre-compute xx and yy.
    % these variables are re-used to achieve faster computation.
    [~,xx,yy] = makegaussian2d(res,2,2,2,2);
    
    socfun = @(dd,wts,c) bsxfun(@minus,dd,c*(dd*wts)).^2 * wts;
    gaufun = @(pp) vflatten(makegaussian2d(res,pp(1),pp(2),pp(3),pp(3),xx,yy,0,0)/(2*pi*pp(3)^2));
    modelfun = @(pp,dd) pp(4)*(socfun(dd,gaufun(pp),restrictrange(pp(6),0,1)).^pp(5));
    
    for idx=1:length(stimuliIdx)
        params = [floor(eval(sprintf('modelTable.pRF_pixel_centers_image_%d_dim0(%d)', stimuliIdx(idx), voxelIdx))),
                  floor(eval(sprintf('modelTable.pRF_pixel_centers_image_%d_dim1(%d)', stimuliIdx(idx), voxelIdx))),
                  eval(sprintf('modelTable.pRF_pixel_sizes_image_%d(%d)', stimuliIdx(idx), voxelIdx)),
                  modelTable.Kay2013_response_gain(voxelIdx),
                  modelTable.Kay2013_output_nonlinearity(voxelIdx),
                  modelTable.Kay2013_SOC_constant(voxelIdx)];
        modelfit = modelfun(params, stimuli(idx,:));
        eval(sprintf('modelTable.MATLAB_predicted_responses_image_%d(%d) = %d;', stimuliIdx(idx),voxelIdx,modelfit));
    end

end